lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua

```diff
@@ -1,54 +1,125 @@
+  / ___/ ___   / // __/  /   | _      __ ____ _ _____ ___     |__ \
+  \__ \ / _ \ / // /_   / /| || | /| / // __ `// ___// _ \    __/ /
+ ___/ //  __// // __/  / ___ || |/ |/ // /_/ // /   /  __/   / __/
+/____/ \___//_//_/    /_/  |_||__/|__/ \__,_//_/    \___/   /____/
 
  Adds functions similarly to regular-e2's self-aware core.
 ]]
 
--- Reused from coroutine-core
-local function getE2UDF(compiler,funcname)
-    local funcs = compiler.funcs
-    local e2func = funcs[funcname]
-    if not e2func then -- We will look for any udfs that have the name before the parenthesis.
-        local fncnameproper = funcname:match("(%w*)%(") or funcname
-        for Name,Runtime in pairs(funcs) do
-            local proper = Name:match("(%w*)%(")
-            if proper and proper==fncnameproper then return Runtime end
-        end
-    else
-        return e2func
-    end
-end
+local isfunction, debug_getinfo = isfunction, debug.getinfo
+local string_find, string_sub, table_GetKeys = string.find, string.sub, table.GetKeys
+local luaTableToE2, getE2UDF, getE2Func = vex.luaTableToE2, vex.getE2UDF, vex.getE2Func
 
-local function getE2Func(compiler,funcname) 
-    local funcs = wire_expression2_funcs
-	local e2func = funcs[funcname]
-    if not e2func then -- We will look for any udfs that have the name before the parenthesis.
-        local fncnameproper = funcname:match("(%w*)%(") or funcname
-		for Name,Data in pairs(funcs) do
-            local proper = Name:match("(%w*)%(")
-            if proper and proper==fncnameproper then return Data[3] end -- Return function runtime
-        end
-    else
-        return e2func[3]
-    end
-end
+-- TODO: Set E2 costs ( __e2setcost(N) ).
 
--- Ex: print(defined("print(...)")) or print(defined("e:health()"))
--- Returns number, 0 being not defined, 1 being defined as an official e2 function, 2 being a user-defined function
+-- Ex: print(defined("print(...)")) or print(defined("health(e:)"))
+-- Returns number, 0 being not defined, 1 being defined as an official e2 function, 2 being a user-defined function.
+-- Note: If you are checking for availability of the builtin function and if you know the signature ahead of time,
+--       then it is preferred to use #ifdef pre-processor statement; this function exists for dynamic kind of checks.
 e2function number defined(string funcname)
-	return getE2Func(self,funcname) and 1 or (getE2UDF(self,funcname) and 2 or 0)
+    -- Check UDF first (see the above note for why).
+    local isUDF, udfDirect = getE2UDF(self, funcname)
+    if udfDirect then return 2 end -- UDF perfect match.
+    local isFunc, funcDirect = getE2Func(self, funcname)
+    if funcDirect then return 1 end -- Builtin perfect match.
+    -- Name only match after this point :(
+    -- Which one to prefer if they are both defined with identical name? It makes sense to prefer UDF?    *smh*
+    -- Should this return a negative number to indicate uncertainity?
```

Comment:
No no...
I think you misunderstand/misread. By the way, you **can** redefine existing E2 functions (only applies to variadic/vararg ones).
Here is a code example to help understand what exactly I meant:
```golo
function void print(Text:string) # Didn't know this? Now you do ;)
{
    print("[info]", Text) # Yes this will call into the builtin print because of the signature/overloading.
}

print(defined("print")) # What do you expect to get from this?
# Since there are 2 of them now, builtin and UDF, both called with the same name.

# More examples:
function applyForce(Ar:array, V:vector)
{
    # Omitted for clarity.
}

print(defined("applyForce"))  # Now again, what do you expect this to say?
```

Reply:
And also of course, you can define your own functions such as `applyForce` (non-variadic) too, as long as the entire UDF signature is *not* already defined.

Reply:
Splitting them doesn't sound bad, just keep the name of 'defined' as it is, it will handle e2functions and 'definedUDF' will be for udfs

Reply:
Could make it overloaded, just like I do for the `udfAll`.
By introducing a number argument (that would be either 0 or 1) to indicate which one to prefer?
So the `defined`'s signature would changed to:
`e2function number defined(string funcname, number preferBuiltin)`
And then a number argument would be checked for (only when it is not certain), do you want me to do this?
Note, changing argument's signature would be a breaking change. (So, how many users are out there using `defined`?)

Reply:
I feel that going with defined vs definedUDF would be a bit more intuitive than making more of these multi-selectional functions, but if you really wanted to do the preferBuiltin route, please keep the number arg optional so it isn't a breaking change like you said (Well it'd still be different in that it returns only e2functions and not udfs)

Reply:
But, hah, rethinking this. You could also use the new `udfNames` / `udfAll` to specifically check for UDF.
So, I believe that the order should be swapped now, it should prefer builtin instead of UDF.